src/strands/agent/agent.py

```
@@ -92,9 +93,34 @@ def __getattr__(self, name: str) -> Callable[..., Any]:
                 A function that when called will execute the named tool.
 
             Raises:
-                AttributeError: If no tool with the given name exists.
+                AttributeError: If no tool with the given name exists or if multiple tools match the given name.
             """
 
+            def find_normalized_tool_name() -> Optional[str]:
+                """Lookup the tool represented by name, replacing characters with underscores as necessary."""
+                tool_registry = self._agent.tool_registry.registry
+
+                if tool_registry.get(name, None):
+                    return name
+
+                # If the desired name contains underscores, it might be a placeholder for characters that can't be
+                # represented as python identifiers but are valid as tool names, such as dashes. In that case, find
+                # all tools that can be represented with the normalized name
+                if "_" in name:
+                    filtered_tools = [
+                        tool_name
+                        for (tool_name, tool) in tool_registry.items()
+                        if "-" in tool_name and tool_name.replace("-", "_") == name
+                    ]
+
+                    if len(filtered_tools) > 1:
+                        raise AttributeError(f"Multiple tools matching '{name}' found: {', '.join(filtered_tools)}")
```

What is the harm of discarding the `"-" in tool_name` condition shouldn't we just be safe doing `tool_name.replace("-", "_") == name`?

But about the strategy as a whole I have a question,

I agree that we should be deterministic if we have multiple tools and fail. But have we considered preventing the tool from being loaded in the first place by just normalizing when we input the tool_name (and then when we retrieve?)



This seems like something that should never happen to me. It's strange to me that multiple tools

---

I considered that - but I wasn't sure how much we wanted to couple direct-method invocation to the tools that agents are allowed to use.  IMHO they're separate use cases, but I can see the argument for keeping them aligned.

---

I personally would rather always have the agent in a state where tools can be executed directly, meaning we'd fail early on ingestion. But, not going to block on that if others disagree, is also something we can add later if needed.

Unrelated, thoughts on the `What is the harm of discarding the` part of the comment above?

---

> Unrelated, thoughts on the What is the harm of discarding the part of the comment above?

I will check; I don't think there's harm, but I don't remember why I added it :)

---

Follow up is in #193